File: src/receipt-capture.ts
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import type { Account } from './account';
import type { Receipt } from './receipt';
import type { ReceiptCapturePlugin, ReqInitialize, ScanType } from './receipt-capture-plugin';
import { LocalNotifications } from '@capacitor/local-notifications';

/**
 * The primary class for interacting with the Plugin.
 */
export class ReceiptCapture {
  private plugin: ReceiptCapturePlugin;

  /**
   * Constructs a new instance of the ReceiptCapture class. Do not construct
   * directly, use {@link instance}.
   * @param plugin - The ReceiptCapturePlugin instance used for communication.
   */
  constructor(plugin: ReceiptCapturePlugin) {
    this.plugin = plugin;
  }

  /**
   * Initializes the receipt capture plugin.
   * 
   * It initializes the Microblink SDK with the License Key and Product Intelligence Key.
   * To use Google OAuth for Gmail, provide a [Google OAuth Client Id]((https://developers.google.com/identity/protocols/oauth2) 
   * 
   * @param licenseKey - The license key for your application.
   * @param productKey - The product intelligence key for your application.
   * @param googleClientId - The Google Client Id for OAuth login. If not provided, Gmail scraping will
   * use IMAP login. 
   * 
   * @throws Error if the initialization fails.
   */
  initialize = async (licenseKey: string, productKey: string, googleClientId: string | undefined): Promise<void> => {
    await LocalNotifications.requestPermissions()
    const notifs = await LocalNotifications.schedule({
      notifications: [
        {
          title: 'Test Notification',
          body: 'Test Notification',
          id: 1,
          schedule: { at: new Date(Date.now() + 1000 * 5) },
          sound: undefined,
          attachments: undefined,
          actionTypeId: '',
          extra: null,
        },
      ],
    });
    console.log('scheduled notifications', notifs);
    
    const req: ReqInitialize = {
      licenseKey: licenseKey,
      productKey: productKey,
      googleClientId: googleClientId,
    };
    await this.plugin.initialize(req).catch((error) => {
      throw Error(error);
    });
  };

  /**
   * Login into a retailer or email account to scan for receipts.
   * 
   * @param username - the username of the account.
   * @param password - the password of the account
   * @param source - the source from that account, that can be an email service or a retailer service.
   * 
   * @returns - the Account interface with the logged in information.
   */
  login = (username: string, password: string, source: string): Promise<Account> =>
    this.plugin.login({ username, password, source });

  /**
   * Log out from one or all {@link Account}.
   * 
   * To logout from all accounts, do not provide username or password. 
   * 
   * @param username - the username of the account that will be logged out.
   * @param password - the password of the account that will be logged out
   * @param source - the source from that account, that can be an email service or a retailer service.
   * 
   * @returns - the Account that logged out
   */
  logout = (username?: string, password?: string, source?: string): Promise<Account> =>
    this.plugin.logout({ username: username!, password: password!, source: source! });

  /**
   * Scan for receipts. 
   * 
   * The scanType method defines which scan will be performed.
   * 
   * @param scanType - The type of the scan.
   * @param account - The account that will be scanned for receipts.
   * 
   * @returns - The scanned Receipt and a boolean indicates the execution.
   */
  scan = (
    scanType: ScanType | undefined,
    account?: Account,
  ): Promise<{ receipt: Receipt; isRunning: boolean; account?: Account }> => this.plugin.scan({ scanType, account });

  /**
   * Retrieves all saved email and accounts.
   * 
   * @returns - an array of Accounts.
   */
  accounts = async (): Promise<Account[]> => await this.plugin.accounts();
}

File: ios/Plugin/Src/Mytiki/Sdk/Capture/Receipt/ReceiptCapture.swift
import Foundation
import Capacitor
import BlinkReceipt

public class ReceiptCapture: NSObject {
    
    var email: Email? = nil
    var physical: Physical? = nil
    var retailer: Retailer? = nil
    
    public static var pendingScanCall: CAPPluginCall?
    
    public func initialize(_ call: CAPPluginCall) {
        let reqInit = ReqInitialize(call)
        let licenseKey = reqInit.licenseKey
        let productKey = reqInit.productKey
        let scanManager = BRScanManager.shared()
        scanManager.licenseKey = licenseKey
        scanManager.prodIntelKey = productKey
        physical = Physical()
        email = Email()
        retailer = Retailer(licenseKey, productKey)
    }
    
   public func login(_ call: CAPPluginCall) {
        let reqLogin = ReqLogin(data: call)
        guard let accountType = AccountCommon.defaults[reqLogin.source] else {
            call.reject("Invalid source: \(reqLogin.source)")
            return
        }
        let account = Account.init(accountType: accountType, user: reqLogin.username, password: reqLogin.password, isVerified: false)
        switch account.accountType.type {
        case .email :
            guard let email = email else {
                call.reject("Call the initialize method.")
                return
            }
            email.login(account, call)
            break
        case .retailer :
            guard let retailer = retailer else {
                call.reject("Call the initialize method.")
                return
            }
            retailer.login(account, call)
            break
        }
    }
    
    public func logout(_ call: CAPPluginCall) {
        let reqLogout = ReqLogin(data: call)
        guard let accountType = AccountCommon.defaults[reqLogout.source] else {
            call.reject("Invalid source: \(reqLogout.source)")
            return
        }
        
        let account = Account.init(accountType: accountType, user: reqLogout.username, password: reqLogout.password, isVerified: false)
        switch account.accountType.type {
        case .email :
            guard let retailer = email else {
                call.reject("Call plugin initialize method.")
                return
            }
            email!.logout(call, account)
            break
        case .retailer :
            guard let retailer = retailer else {
                call.reject("Call plugin initialize method.")
                return
            }
            retailer.logout(account, call)
            break
        }
    }
    
    public func accounts(_ call: CAPPluginCall) {
        guard let retailer = retailer else {
            call.reject("Call plugin initialize method.")
            return
        }
        
        let data = retailer.accounts(call)
        print(data)
    }
    
    public func scan(_ call: CAPPluginCall) {


        let req = ReqScan(data: call)
        if req.account == nil {
            switch req.scanType {
            case .EMAIL:
                Email().scan(call, req.account)
            case .RETAILER:
                Retailer().orders(req.account, call)
            case .PHYSICAL:
                guard let physical = physical else {
                call.reject("Call plugin initialize method.")
                return
            }
            ReceiptCapture.pendingScanCall = call
            physical.scan()
            case .ONLINE:
                Email().scan(call, req.account)
                Retailer().orders(req.account, call)
            default:
                call.reject("invalid scan type for account")
            }
        } else {
            switch req.scanType {
            case .EMAIL:
                Email().scan(call, req.account)
            case .RETAILER:
                Retailer("","").orders(req.account, call)
            default:
                call.reject("invalid scan type for account")
            }
        }

    }
    
}
